src/ModelContextProtocol/Authentication/AuthorizationDelegatingHandler.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            return await HandleUnauthorizedResponseAsync(request, response, cancellationToken).ConfigureAwait(false);
+        }
+
+        return response;
+    }
+
+    /// <summary>
+    /// Handles a 401 Unauthorized response by attempting to authenticate and retry the request.
+    /// </summary>
+    private async Task<HttpResponseMessage> HandleUnauthorizedResponseAsync(
+        HttpRequestMessage originalRequest, 
+        HttpResponseMessage response, 
+        CancellationToken cancellationToken)
+    {
+        // Gather the schemes the server wants us to use from WWW-Authenticate headers
+        var serverSchemes = ExtractServerSupportedSchemes(response);
+
+        // Find the intersection between what the server supports and what our provider supports
+        string? bestSchemeMatch = null;
+
+        // First try to find a direct match with the current scheme if it's still valid
+        string schemeUsed = originalRequest.Headers.Authorization?.Scheme ?? _currentScheme ?? string.Empty;
+        if (!string.IsNullOrEmpty(schemeUsed) && 
+            serverSchemes.Contains(schemeUsed) && 
+            _credentialProvider.SupportedSchemes.Contains(schemeUsed))
+        {
+            bestSchemeMatch = schemeUsed;
+        }
+        else
+        {
+            // Find the first server scheme that's in our supported set
+            bestSchemeMatch = serverSchemes.Intersect(_credentialProvider.SupportedSchemes, StringComparer.OrdinalIgnoreCase).FirstOrDefault();
+
+            // If no match was found, either throw an exception or use default
+            if (bestSchemeMatch is null)
+            {
+                if (serverSchemes.Count > 0)
+                {
+                    throw new InvalidOperationException(
+                        $"No matching authentication scheme found. Server supports: [{string.Join(", ", serverSchemes)}], " +
+                        $"Provider supports: [{string.Join(", ", _credentialProvider.SupportedSchemes)}].");
+                }
+
+                // If the server didn't specify any schemes, use the provider's default
+                bestSchemeMatch = _credentialProvider.SupportedSchemes.FirstOrDefault();
+            }
+        }
+
+        // If we have a scheme to try, use it
+        if (bestSchemeMatch != null)
+        {
+            // Try to handle the 401 response with the selected scheme
+            var (handled, recommendedScheme) = await _credentialProvider.HandleUnauthorizedResponseAsync(
+                response,
+                bestSchemeMatch,
+                cancellationToken).ConfigureAwait(false);
+
+            if (!handled)
+            {
+                throw new McpException(
+                    $"Failed to handle unauthorized response with scheme '{bestSchemeMatch}'. " +
+                    "The authentication provider was unable to process the authentication challenge.");
+            }
```

Comment:
I don't love having a handled/Success bool and then just immediately rethrowing. Looking at the BasicOAuthProvider, it just logs any exceptions to the console and returns false. Bubbling up the original exception to the caller would make things much easier to diagnose.

I don't mind wrapping any exceptions thrown by HandleUnauthorizedResponseAsync with another exception containing this message, but the only reason to use something like handled in C# is if there's a chance you don't need to throw at all.